### PR TITLE
Inconsistent handling of files in an output folder - full vs. incremental build

### LINKED ISSUE #2920: Inconsistent handling of files in an output folder - full vs. incremental build
Observed in Xtext 2.28.0:

I was debugging an issue where a search using resource descriptions produced unexpected results, because we encountered a mix of descriptions from `src` and `bin` folders.

Disclaimer: against prior advice we indeed use index search during building, after taking appropriate steps to ensure a complete index at the point of usage.

I found that the problem does not occur during **full builds**, because `ToBeBuiltComputer.updateProject()` rejects any `IFolder` that is an output folder (using `isHandled(IFolder)` etc.).

During **incremental build**, however, `XtextBuilder.createDeltaVisitor()` creates a visitor that does not have a branch for `IFolder` and hence we fall through to `return true;` causing all files in output folders to be visited, too.

This looks inconsistent to me.

We are currently fixing this by rolling a custom XtextBuilder, but I wonder if perhaps there is some reasoning behind this difference between full & incremental builds? If not, adding a corresponding branch in said delta visitor might actually bring a performance benefit for everybody, no? :)

I should add, that I don't see an easy way to extract our situation into a sharable test...

Browsing previous issues, that closest I could find is #2512 which looks inconclusive to me.

### PR DESCRIPTION
**Pull request by stephan-herrmann**
fixes #2920

### CODE REVIEW

**Comment by stephan-herrmann** [comment]
I did poke around existing tests, to see if anything could be used as a template for creating a test challenging this change, but I failed to find any test that would even invoke `XtextBuilder.incrementalBuild()`. If such tests exist, I'd be happy to try to work from there. Any hints?

**Comment by cdietrich** [comment]
i assume some of the xtend.ide tests test some of the usecases
/org.eclipse.xtend.ide.tests/src/org/eclipse/xtend/ide/tests/builder
as these do creation/changing files

=> you might check the history of code in xtext-eclipse
and then search corrsponding tests
in xtext-xtend
e.g. https://github.com/eclipse/xtext-xtend/pull/1309/files

there is also a builder test language in builder.tests.
but i am not sure what it is used for

**Comment by stephan-herrmann** [comment]
By copying from `org.eclipse.xtext.builder.impl.Bug386476Test` I managed to write a test that get's close to the problem. Just when I thought I have a witness, I learned that the problem is normally prevented deep down in the following call chain:
* XtextBuilder.createDeltaVisitor().new IResourceDeltaVisitor().visit()
   * toBeBuiltComputer.updateStorage(..) *<-- here we indeed pass a resource from "bin"*
      * ToBeBuildComputer.getURI(IStorage)
         * mapper.getUri(file)
            * isValidUri(..)
               * uriValidator.isValid(..)
                  * resourceServiceProvider.canHandle(..)

In the last call, if the provider is a `DefaultResourceUIServiceProvider`, then this snippet blocks the URI:
```
	if (isJavaCoreAvailable()) {
		return !isJavaTargetFolder(storage);
	}
```

Going back to our original problematic case, I realize that we have a test case with a combination that's perhaps unsupported: we invoke the builder on a proper **workspace**, but we **don't have UI**, so we only get a `DefaultResourceServiceProvider`.  This situation arose when I tried to create minimal tests with capability to challenge the builder.

Am I right that Xtext doesn't have a "mode" for a headless eclipse? Is that why only the "UI" variant asks `isJavaCoreAvailable()`? IOW, no distinction headless vs. ui but only standalone vs. IDE?

If the answer is indeed "unsupported", then my change may be "correct but irrelevant". So would you still want it?

@szarnekow  wdyt?

**Comment by cdietrich** [comment]
My assumption would be that even in headless eclipse with workspace
DefaultResourceUIServiceProvider Is bound.
As Xtext does not support this ootb
At some customers I saw special activators / ui modules for that use case

how do you implement that

**Comment by LorenzoBettini** [comment]
@stephan-herrmann in such tests, do you use the UI injector provider?

**Comment by stephan-herrmann** [comment]
Our headless tests use a slim handcrafted `HeadlessSharedStateModule` including a `HeadlessJdtHelper`.

All this is inspired by https://github.com/eclipse-jdt/eclipse.jdt.core/tree/master/org.eclipse.jdt.core.tests.builder which has no dependencies on anything ui :)

DefaultResourceUIServiceProvider, OTOH, lives in org.eclipse.xtext.ui and hence will never work in a truly headless eclipse, where none of org.eclipse.ui* are available.

Initially I had assumed the separation between o.e.xtext and o.e.xtext.ui can be compared to the separation between o.e.jdt.core and o.e.jdt.ui, but apparently xtext is not prepared to leverage jdt.core without any ui.

As all this is only about minimizing tests (we don't ship headless applications), it is not a problem for production.

**Comment by stephan-herrmann** [comment]
> if all the deps are optional

but that's not the case :smile:  - I see lots of non-optional ui dependencies in xtext.ui, so this plugin will simply not resolve in a truly headless environment.

> xtext.ui still should live in an equinox env without workbench. of course we dont have separate bundles though

fair enough.

**Comment by cdietrich** [comment]
the funny thing is in the project i worked with andrey loskutov on it does. (of course with a bunch of extra bundles in the product)

**Comment by stephan-herrmann** [comment]
With all that I learned in this PR, I'd be fine with closing it without merge.

Is anyone still planing to review? I mean it still seems to be the "right" thing to do, but the effective benefit might be low (a tiny performance improvement, more symmetric code).

File: org.eclipse.xtext.builder.tests/src/org/eclipse/xtext/builder/impl/GH2920Test.java [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+	@Before
+	public void createProjectUnderTest() throws Exception {
+		javaProject = createJavaProject("NoResourceFromOutput");
+		addNature(javaProject.getProject(), XtextProjectHelper.NATURE_ID);
+		build();
+	}
+
+	@Override
+	public void build(IBuildContext context, IProgressMonitor monitor) throws CoreException {
```
[the remark this answers is not in the document]
Comment (stephan-herrmann) [reply org.eclipse.xtext.builder.tests/src/org/eclipse/xtext/builder/impl/GH2920Test.java:44]: Have you seen [this comment](https://github.com/eclipse/xtext/pull/2921#issuecomment-1918044330)?

In essence we'd need an environment *with* `XtextBuilder` but *without* `DefaultResourceUIServiceProvider`. 

I can confirm that one of our tests, which uses `DefaultResourceServiceProvider` instead, does trigger the problem. Do you want me to construct such a test for xtext? It looked like that would need jumping through some hoops. 

That's why I was uncertain whether we should actually pursue this further...

**Comment by LorenzoBettini** [comment]
@stephan-herrmann maybe I didn't get the overall picture but if the problem is



> In the last call, if the provider is a `DefaultResourceUIServiceProvider`, then this snippet blocks the URI:
> 
> ```
> 	if (isJavaCoreAvailable()) {
> 		return !isJavaTargetFolder(storage);
> 	}
> ```

To recreate your scenario, where you wouldn't want the `isJavaCoreAvailable` to return true even in a UI test, shouldn't it be enough to create a custom implementation of `JdtHelper` that implements https://github.com/eclipse/xtext/blob/88acf1555988cd81e1756abc0d6d2c3f1f1efa11/org.eclipse.xtext.ui.shared/src/org/eclipse/xtext/ui/shared/JdtHelper.java#L52 always returning false and make sure your test is run with your custom implementation injected (this requires to create a custom injector provider)?

could that recreate the problem and verify the fix correctly, @szarnekow ?

**Comment by stephan-herrmann** [comment]
> @stephan-herrmann maybe I didn't get the overall picture but if the problem is
> 
> > In the last call, if the provider is a `DefaultResourceUIServiceProvider`, then this snippet blocks the URI:
> > ```
> > 	if (isJavaCoreAvailable()) {
> > 		return !isJavaTargetFolder(storage);
> > 	}
> > ```
> 
> To recreate your scenario, where you wouldn't want the `isJavaCoreAvailable` to return true even in a UI test, shouldn't it be enough to create a custom implementation of `JdtHelper` that implements ...

This is correct, but I don't know if it's significantly easier to re-bind JdtHelper then IResourceServiceProvider. Frankly, I just didn't feel like creating a whole now language with its own guice module, assuming that would be overkill.

**Comment by LorenzoBettini** [comment]
I didn't mean create a new language: just use a custom ui injector provider, inheriting the one currently used. If you want I can have a try.

**Comment by stephan-herrmann** [comment]
> I didn't mean create a new language: just use a custom ui injector provider, inheriting the one currently used. If you want I can have a try.

that'd be cool, thanks.

**Comment by LorenzoBettini** [comment]
Apparently, I was too optimistic because I've never had problems in customizing test injectors for headless tests, but I never did that for UI tests.

This does not work as expected:

```java
@RunWith(XtextRunner.class)
@InjectWith(GH2920Test.BuilderTestLanguageInjectorProviderCustom.class)
public class GH2920Test extends AbstractParticipatingBuilderTest {
	private IJavaProject javaProject;

	@Inject private IResourceDescriptionsProvider descriptionsProvider;

	@Singleton
	public static class JdtHelperCustom extends JdtHelper {
		@Override
		protected boolean computeJavaCoreAvailable() {
			return false;
		}
	}

	public static class BuilderTestLanguageInjectorProviderCustom extends BuilderTestLanguageInjectorProvider {
		@Override
		public Injector getInjector() {
			Module mixin = Modules2.mixin(new BuilderTestLanguageRuntimeModule(),
				new SharedStateModule() {
					@Override
					public Provider<IJdtHelper> provideJdtHelper() {
						return new Provider<>() {
							@Override
							public IJdtHelper get() {
								return new JdtHelperCustom();
							}
						};
					}
				},
				new BuilderTestLanguageUiModule(TestsActivator.getInstance())
			);
			return Guice.createInjector(mixin);
		}
	}
...
```

In fact, the custom injector is used only to inject things in the test case, but the workspace and other services, including the build participants, are injected with the executable extension factory, i.e., with the injector created in the test language activator, which cannot be customized.

Maybe @szarnekow can shed some light on whether what I wanted to do is feasible without creating a new language.

**Comment by LorenzoBettini** [comment]
I think I found a solution to recreate the problem. I have to do a few clean up and then I'll tell you about that

**Comment by LorenzoBettini** [comment]
@stephan-herrmann I created a PR against your branch with the test reproducing the problem. Of course, in the PR the test succeeds, but if you try uncommenting your modifications to `XtextBuilder` it properly fails.
My PR is based on top your branch https://github.com/stephan-herrmann/xtext/tree/issue2920 so it also contains your fix for the issue.
If you merge it in your branch, this very PR will be updated.

**Comment by stephan-herrmann** [comment]
> @stephan-herrmann I created a PR against your branch with the test reproducing the problem. Of course, in the PR the test succeeds, but if you try uncommenting your modifications to `XtextBuilder` it properly fails. My PR is based on top your branch https://github.com/stephan-herrmann/xtext/tree/issue2920 so it also contains your fix for the issue. If you merge it in your branch, this very PR will be updated.

Thanks, @LorenzoBettini for going the extra mile!

Now on top of your changes I proposed a variant that comes even closer to our original problem: in our case the IResourceServiceProvider was the point where behavior deviated, so I bound something that looks like an IResourceUIServiceProvider (because that's what is requested in the test setup) and behaves like a DefaultResourceServiceProvider. This, too, succeeds to demonstrate the problem when the actual fix is not in.

If that's too much magic in the test, I'm fine with reverting [this last commit](https://github.com/eclipse/xtext/pull/2921/commits/4b43a2103fbd3775c4bde5f46dfff5cbe53dc486).

**Review by szarnekow — APPROVED** [review]
Very nice!

**Comment by szarnekow** [comment]
Thank you for the improved test. I like the approach! Let’s wait for the build to succeed and merge afterwards.

**Pull request merged**